### PR TITLE
Add RPC InvokeContractVerify

### PR DESCRIPTION
**Pull request by chenzhitong**
close https://github.com/neo-project/neo/issues/1977

### CODE REVIEW

**Comment by chenzhitong** [comment]
Blocked by this problem https://github.com/neo-project/neo-devpack-dotnet/issues/397

**Comment by chenzhitong** [comment]
Blocked by this problem https://github.com/neo-project/neo-node/issues/684

**Comment by chenzhitong** [comment]
**request**

```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "invokecontractverify",
    "params": [
        "0x63c7e30a5eee6c8208f9ba073b3424e8b592c26c"
    ]
}
```

**Response**

```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "script": "IQwU9mRDSY04eNMrmU5OEoPGk0Qh2v4hQfgn7IxAEUBXAQAhDAVIZWxsbyFBm/ZnzkGSXegxcCMFAAAAaEBXAQEheHBoJwoAAAAjIAAAAAwFSGVsbG8MBVdvcmxkIVBBm/ZnzkHmPxiEIUBXAQIhITWQ////ELNwaCcaAAAADBFObyBhdXRob3JpemF0aW9uLiE6eHkhUEExxjMdIUBXAQAhITVd////ELNwaCcaAAAADBFObyBhdXRob3JpemF0aW9uLiE6IUHGnx3wIUAhVgEMFPZkQ0mNOHjTK5lOThKDxpNEIdr+YCFA",
        "state": "HALT",
        "gasconsumed": "0.0101685",
        "exception": null,
        "stack": [
            {
                "type": "Integer",
                "value": "1"
            },
            {
                "type": "Integer",
                "value": "1"
            }
        ]
    }
}
```

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 1]
```diff
@@ -107,6 +107,45 @@ private static Signers SignersFromJson(JArray _params)
...
+                ProcessInvokeWithWallet(json);
+            }
+            return json;
+        }
+
+        [RpcMethod]
+        protected virtual JObject InvokeContractVerify(JArray _params)
+        {
+            UInt160 script_hash = UInt160.Parse(_params[0].AsString());
+            ContractParameter[] args = _params.Count >= 3 ? ((JArray)_params[1]).Select(p => ContractParameter.FromJson(p)).ToArray() : new ContractParameter[0];
```
Comment (shargon) [inline src/RpcServer/RpcServer.SmartContract.cs:145]:
```suggestion
            ContractParameter[] args = _params.Count >= 2 ? ((JArray)_params[1]).Select(p => ContractParameter.FromJson(p)).ToArray() : new ContractParameter[0];
```

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 2]
```diff
@@ -107,6 +107,45 @@ private static Signers SignersFromJson(JArray _params)
             return ret;
         }
 
+        private JObject GetVerificationResult(UInt160 scriptHash, ContractParameter[] args)
+        {
+            var snapshot = Blockchain.Singleton.GetSnapshot();
+            var contract = snapshot.Contracts.TryGet(scriptHash);
+            var methodName = "verify";
+
+            using ApplicationEngine engine = ApplicationEngine.Create(TriggerType.Verification, new Transaction(), snapshot.Clone());
```
Comment (shargon) [inline src/RpcServer/RpcServer.SmartContract.cs:116]: I think that we don't need an empty object, because it will fault like `null` if it's needed.

Comment (chenzhitong) [reply src/RpcServer/RpcServer.SmartContract.cs:116]: Add a parameter that allows the user to customize the Transaction?

Comment (shargon) [reply src/RpcServer/RpcServer.SmartContract.cs:116]: At least the signers, `verify` "always" check the `Signers`

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 3]
```diff
@@ -107,6 +107,53 @@ private static Signers SignersFromJson(JArray _params)
             return ret;
         }
 
+        private JObject GetVerificationResult(UInt160 scriptHash, ContractParameter[] args, Signers signers = null)
+        {
+            var snapshot = Blockchain.Singleton.GetSnapshot();
+            var contract = snapshot.Contracts.TryGet(scriptHash);
```
Comment (shargon) [inline src/RpcServer/RpcServer.SmartContract.cs:113]: If it's null, we can throw an error now

Comment (chenzhitong) [reply src/RpcServer/RpcServer.SmartContract.cs:113]: I think signers should be allowed to be null, and some contract Verify methods don't necessarily require a signature.

**Review by shargon — APPROVED** [review]

**Comment by devhawk** [comment]
This PR is not compiling due to use of snapshot.Contracts in GetVerificationResults

**Pull request merged**